### PR TITLE
Disallow multiple parent-join fields per mapping

### PR DESCRIPTION
**Pull request by jimczi**
This change ensures that there is a single parent-join field defined per mapping.
The verification is done through the addition of a special field mapper (MetaJoinFieldMapper) with a unique name (_parent_join) that is registered to the mapping service
when the first parent-join field is defined. If a new parent-join is added, this field mapper will clash with the new one and the update will fail.
This change also simplifies the parent join fetch sub phase by retrieving the parent-join field without iterating on all fields in the mapping.

### CODE REVIEW

**Comment by jimczi** [comment]
\cc @jpountz

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java [hunk 1]
```diff
@@ -69,11 +73,21 @@
         }
     }
 
-    static String getParentIdFieldName(String joinFieldName, String parentName) {
+    /**
+     * Returns the {@link ParentJoinFieldMapper} associated with the <code>service</code> or null
+     * if there is no parent-join field in this mapping.
+     */
+    public static ParentJoinFieldMapper getMapper(MapperService service) {
```
Comment (martijnvg) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java]: :+1: (I added the same method here locally (differently implemented) )

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java [hunk 2]
```diff
@@ -214,14 +223,18 @@ public Object valueForDisplay(Object value) {
...
     private List<ParentIdFieldMapper> parentIdFields;
 
     protected ParentJoinFieldMapper(String simpleName,
                                     MappedFieldType fieldType,
                                     Settings indexSettings,
+                                    MetaJoinFieldMapper uniqueFieldMapper,
                                     List<ParentIdFieldMapper> parentIdFields) {
         super(simpleName, fieldType, Defaults.FIELD_TYPE, indexSettings, MultiFields.empty(), null);
         this.parentIdFields = parentIdFields;
+        this.uniqueFieldMapper = uniqueFieldMapper;
```
Comment (martijnvg) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java]: should we also add `uniqueFieldMapper.setFieldMapper(this)` here ?

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java [hunk 3]
```diff
@@ -241,7 +254,9 @@ public JoinFieldType fieldType() {
 
     @Override
     public Iterator<Mapper> iterator() {
-        return parentIdFields.stream().map((field) -> (Mapper) field).iterator();
+        List<Mapper> mappers = new ArrayList<> (parentIdFields);
+        mappers.add(uniqueFieldMapper);
+        return mappers.iterator();//parentIdFields.stream().map((field) -> (Mapper) field).iterator();
```
Comment (martijnvg) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java]: remove comment?

**Review by martijnvg — APPROVED** [review]
Left a question and a small comment. LGTM

File: modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java (hunk 2 above)
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentJoinFieldMapper.java]: yes, thanks

**Review by jpountz — COMMENTED** [review]
I think this is an ok hack for now, let's just make sure we revisit later, eg. the error message could be better.

**Comment by jimczi** [comment]
Thanks @jpountz and @martijnvg 

> let's just make sure we revisit later, eg. the error message could be better.

Yep, we also need to state clearly in the docs why it is forbidden.

**Comment by pangyikhei** [comment]
If I'm interpreting the statement in the blog https://www.elastic.co/blog/this-week-in-elasticsearch-and-apache-lucene-2017-06-05 correctly:

>  Good progress has been made in 5.5 with the introduction of the ParentJoinFieldMapper, a field mapper that creates parent/child relation within documents of the same index. Only one join-field is allowed per index. 

Does this mean we can't have a parent with multiple child types anymore?
Ex. **Question** type as the parent of **Answer** and **Comment** child types.

**Comment by martijnvg** [comment]
No, it just means one join field can be configured per index. However a join field can be configured with multiple relations like you described above.

**Pull request merged**